File: packages/fastapi/src/fastapi_quanttide_finance/models/classification_result.py
from datetime import datetime

from sqlalchemy import Boolean, Column, DateTime, Float, ForeignKey, Integer, String, func
from sqlalchemy.dialects.sqlite import JSON

from fastapi_quanttide_finance.database import Base


class ClassificationResult(Base):
    __tablename__ = "classification_result"

    id = Column(Integer, primary_key=True, autoincrement=True)
    normalized_record_id = Column(
        Integer, ForeignKey("normalized_record.id"), nullable=False
    )
    taxonomy = Column(String(50), nullable=False)
    category = Column(String(255), nullable=False)
    tags = Column(JSON, nullable=True)
    classifier_kind = Column(String(50), nullable=False)
    confidence = Column(Float, nullable=True)
    model_version = Column(String(50), nullable=True)
    review_status = Column(String(50), nullable=False, default="candidate")
    is_active = Column(Boolean, nullable=False, default=True)
    created_at = Column(DateTime, nullable=False, server_default=func.now())
    updated_at = Column(
        DateTime,
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

File: packages/finance/dart/lib/src/dto/classification_result.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'enums.dart';

part 'classification_result.freezed.dart';
part 'classification_result.g.dart';

@freezed
class ClassificationResultDto with _$ClassificationResultDto {
  const factory ClassificationResultDto({
    required int id,
    @JsonKey(name: 'normalized_record_id') required int normalizedRecordId,
    @Default('expense_type') String taxonomy,
    required String category,
    @JsonKey(name: 'tags') Map<String, dynamic>? tags,
    @JsonKey(
      name: 'classifier_kind',
      unknownEnumValue: ClassifierKind.unknown,
    )
    required ClassifierKind classifierKind,
    @JsonKey(name: 'confidence') double? confidence,
    @JsonKey(name: 'model_version') String? modelVersion,
    @JsonKey(
      name: 'review_status',
      unknownEnumValue: ReviewStatus.unknown,
    )
    @Default(ReviewStatus.candidate)
    ReviewStatus reviewStatus,
    @JsonKey(name: 'is_active') @Default(true) bool isActive,
    @JsonKey(name: 'created_at') required DateTime createdAt,
    @JsonKey(name: 'updated_at') required DateTime updatedAt,
  }) = _ClassificationResultDto;

  factory ClassificationResultDto.fromJson(Map<String, dynamic> json) =>
      _$ClassificationResultDtoFromJson(json);
}
